pkg/provisioner/ironic/hardwaredetails/hardwaredetails.go

```diff
@@ -95,12 +95,29 @@ func getNICDetails(ifdata []introspection.InterfaceType,
 	return nics
 }
 
+func getDiskType(diskdata introspection.RootDiskType) metal3v1alpha1.DiskType {
+	if diskdata.Name == "" {
+		return metal3v1alpha1.UNKNOWN
```

Comment:
I don't know that we'll ever be able to remove the Rotational field when this discards some of that data in some circumstances. Wouldn't it be better to just fall back to returning "SSD" or "HDD" in this case?

Reply:
There is no possible way that name field could be empty, name filed here is kernel device name, "/dev/sda" for example.
If this name is empty device shouldn't exists in first place.

We be better to remove omitempty from struct field, to make this filed mandatory.

Also there is no way of knowing what type of device is this "SSD" or "HDD", how can we chose what default value is for this (for none existent device).

Defaulting to either one of them will look strange, none existent device of some type?

But when using unknown, this effectively sets disk type to empty string and this is exactly default value for uninitialized string in golang, this makes more sense when combined with empty kernel name

Reply:
> There is no possible way that name field could be empty

Then why do we have an enum value for something that can't happen?

Reply:
> Then why do we have an enum value for something that can't happen?
This could happen if IPA has some bug and allowed by struct field to happen.

getDiskType() is expected to return something in this case it is empty string but with type aliasing, hence enum for empty string

Reply:
You said it can't happen so my question was why, of all the infinite things that can't happen, do we choose to pretend to handle this one as if it could?

Reply:
Can't happen on normal working linux environment, linux device needs to have name.

Can happen when IPA sends malformed data, current CRD validation rules allow that.

Reply:
So device in terms of CRD validation can have any data set, can have Vendor, Model, SizeBytes, ... set but have empty Name.

(IPA uses lsblk output to get this data parsed)

Knowing all of this, as end user I can't grasp what to do with disk data that has all of fields set except name, how it was set, from where this data comes, how to match this data to actual disk inside linux.

Reply:
I think you're focusing too much on the CRD validation. The HardwareDetails can only be set by the BMO, and they're set by this code based on the the output of IPA. The CRD validation is irrelevant except insofar as it must allow all of the values that IPA can send. We don't need to worry about values that IPA can't send, regardless of whether they would be allowed by the OpenAPI specification. We should treat the OpenAPI schema as documentation for the user on what values to expect.

Reply:
Still I do not know how to choose what type to set for a case when name is empty.

```
if diskdata.Name == "" {...}
```

How about just removing this case altogether?

Reply:
Yes, I think we should just remove this check (and the UNKNOWN enum value) since it can't happen.

Reply:
done, thanks